src/sources/prometheus/scrape.rs

```diff
@@ -529,6 +565,101 @@ mod test {
         }
     }
 
+    #[tokio::test]
+    async fn test_prometheus_request_query() {
+        let in_addr = next_addr();
+
+        let dummy_endpoint = warp::path!("metrics").and(warp::query::raw()).map(|query| {
+            format!(
+                r#"
+                    promhttp_metric_handler_requests_total{{query="{}"}} 100 1612411516789
+                "#,
+                query
+            )
+        });
+
+        tokio::spawn(warp::serve(dummy_endpoint).run(in_addr));
+
+        let query = HashMap::from([
+            ("key1".to_string(), vec!["val2".to_string()]),
+            (
+                "key2".to_string(),
+                vec!["val1".to_string(), "val2".to_string()],
+            ),
+        ]);
+
+        let config = PrometheusScrapeConfig {
+            endpoints: vec![format!("http://{}/metrics?key1=val1", in_addr)],
+            scrape_interval_secs: 1,
+            instance_tag: Some("instance".to_string()),
+            endpoint_tag: Some("endpoint".to_string()),
+            honor_labels: false,
+            query: Some(query.clone()),
+            auth: None,
+            tls: None,
+        };
+
+        let (tx, rx) = SourceSender::new_test();
+        let source = config
+            .build(SourceContext::new_test(tx, None))
+            .await
+            .unwrap();
+
+        tokio::spawn(source);
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let events = test_util::collect_ready(rx).await;
+        assert!(!events.is_empty());
+
+        let metrics: Vec<_> = events
+            .into_iter()
+            .map(|event| event.into_metric())
+            .collect();
+
+        let expected = HashMap::from([
+            (
+                "key1".to_string(),
+                vec!["val1".to_string(), "val2".to_string()],
+            ),
+            (
+                "key2".to_string(),
+                vec!["val1".to_string(), "val2".to_string()],
+            ),
+        ]);
+
+        for metric in metrics {
+            // note: keys in HashMaps are in undefined order, therefore a
+            // pair of maps must be compared back and forth key-by-key
+            if let Some(q) = metric.tag_value("query") {
```

Comment:
```suggestion
            let q = metric.tag_value("query").expect("query must be tagged")
```

Would avoid bugs that result in `query` not being tagged.

Reply:
Ah good idea. It also simplifies the code more. Done!